mongomock/store.py

```diff
@@ -78,29 +82,97 @@ def is_created(self):
     def drop(self):
         self._documents = collections.OrderedDict()
         self.indexes = {}
+        self._ttl_indexes = {}
         self._is_force_created = False
 
+    def create_index(self, index_name, index_dict):
+        self.indexes[index_name] = index_dict
+        if index_dict.get('expireAfterSeconds') is not None:
+            self._ttl_indexes[index_name] = index_dict
+
+    def drop_index(self, index_name):
+        # The main index object should raise a KeyError, but the
+        # TTL indexes have no meaning to the outside.
+        del self.indexes[index_name]
+        self._ttl_indexes.pop(index_name, None)
+
     @property
     def is_empty(self):
+        self._remove_expired_documents()
         return not self._documents
 
     def __contains__(self, key):
+        self._remove_expired_documents()
         return key in self._documents
 
     def __getitem__(self, key):
+        self._remove_expired_documents()
         return self._documents[key]
 
     def __setitem__(self, key, val):
         with lock:
             self._documents[key] = val
 
     def __delitem__(self, key):
-        del self._documents[key]
+        self._documents.pop(key)
 
     def __len__(self):
+        self._remove_expired_documents()
         return len(self._documents)
 
     @property
     def documents(self):
+        self._remove_expired_documents()
         for doc in six.itervalues(self._documents):
             yield doc
+
+    def _remove_expired_documents(self):
+        for index in six.itervalues(self._ttl_indexes):
+            self._expire_documents(index)
+
+    def _expire_documents(self, index):
+        # TODO(juannyg): use a caching mechanism to avoid re-expiring the documents if
+        # we just did and no document was added / updated
+
+        # Ignore non-integer values
+        try:
+            expiry = int(index['expireAfterSeconds'])
+        except ValueError:
+            return
+
+        # Ignore commpound keys
+        if len(index['key']) > 1:
+            return
+
+        # "key" structure = list of (field name, direction) tuples
+        ttl_field_name = index['key'][0][0]
+        ttl_now = mongomock.utcnow()
```

Comment:
Here as well, import it from helpers please.

Reply:
I can do this but it has an interesting side-effect - [the test specific to mocking will fail](https://github.com/juannyG/mongomock/blob/ttl-index/tests/test__collection_api.py#L1389) because the "usage path" has changed.

There are two ways to go about this - the first I think will be less appealing, because I believe it goes against your original intent.

In `monogmock/store.py`

    from mongomock import helpers
    ...
    ttl_now = helpers.utcnow()

the test patch would then become

    with mock.patch('mongomock.store.helpers.utcnow') as mongomock_utcnow:

The implication of this is that now users would have to be aware of the different usages/locations in the library to target appropriately.

The alternative would allow for centralized mock that would affect all usages in the library that follow the same usage-pattern:

In `monogmock/store.py`

    import mongomock.helpers
    ...
    ttl_now = mongomock.helpers.utcnow()

the test patch would then become

    with mock.patch('mongomock.helpers.utcnow') as mongomock_utcnow:

From your earlier comments in the issue I originally opened, it seems as though the latter is the preferred approach: as long as anywhere in the library we want to allow mocking access uses `helpers.utcnow` in the same way - users would only need to mock in one place for "global" `utcnow` manipulation.

Subtle - but I wanted to raise this and get your input before I made any more changes. Thx!

Reply:
I see your point, and you're entirely right. Let's just use `mongomock.utcnow` in our code as well. I'll try to make sure we don't use the `mongomock` import internally for other reasons. Maybe add a comment just above the `import mongmomock` explaining why we do it like that? Maybe just redirect readers to the README (no need for a huge explanation).